### ISSUE TITLE
Allow authoring commit message prior to invoking git-commit

### ISSUE THREAD

**Issue by jwiegley** [magit/magit#3693]
At present, when typing `c c`, it invokes Git and waits for it connect back to Emacs using `emacsclient`.

For projects that don't use a `pre-commit` hook, it would be much faster to ask for the commit message in a buffer, and then simply pass that message to `git-commit` using `-m`.

I'd like this feature because, for reasons I haven't been able to track down yet, `emacsclient` stops working with my Emacs after a few hours, necessitates several restarts a day for commits to work. It's also very, very slow for the commit message window to open up.

**Comment by tarsius** [comment]
> For projects that don't use a `pre-commit` hook, it would be much faster to ask for the commit message in a buffer, and then simply pass that message to `git-commit` using `-m`.

But even in that case doing the latter would lose functionality. For example when a rebase stops, then it puts the message of `HEAD` into `git/COMMIT_EDITMSG`. `git commit` also puts useful comments into the file, such as "these are the files that have been modified". You wouldn't get to see that with `git commit -m`.

> I'd like this feature because, for reasons I haven't been able to track down yet, `emacsclient` stops working with my Emacs after a few hours, necessitates several restarts a day for commits to work.

Is that a regression in Emacs 26.2 and/or 27 or do you get that with 26.1 too? Regardless, I think you should investigate this issue instead of looking for workarounds.

> It's also very, very slow for the commit message window to open up.

It shouldn't be and for most users this works very well now (and has been for a few years). Are you sure the time isn't spend generating the diff (which can be aborted using `C-g`)?

Assuming it works, the current approach is far superior to `git commit -m`. I don't intend to bring back the latter. I don't have the time for that, it would complicate matters, and if you didn't have the issue that you are having, then you probably wouldn't ask for this.

In summary, I think time is better spend investigating the issue than implementing a workaround.

**Comment by tarsius** [comment]
Also a properly working `emacsclient` is required for rebasing and in this case it is not possible to implement it differently. `git rebase` needs to use the `$EDITOR`. (Well, one could reimplement `git rebase` in elisp I suppose.)

Actually that is not quite true. `with-editor` does not necessarily need a working `emacsclient`/`server.el`. It also implements an alternative that uses stdin/stdout and is intended for use over Tramp. But it can also be used locally, see `with-editor-emacsclient-executable`, i.e. use the "nuclear option" as mentioned at https://magit.vc/manual/with-editor/Configuring-With_002dEditor.html.
